instrumentation/kotlinx-coroutines/kotlinx-coroutines-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kotlinxcoroutines/instrumentationannotations/AnnotationInstrumentationModule.java

```diff
@@ -50,4 +47,13 @@ public ElementMatcher.Junction<ClassLoader> classLoaderMatcher() {
   public List<TypeInstrumentation> typeInstrumentations() {
     return singletonList(new WithSpanInstrumentation());
   }
+
+  @Override
+  public List<String> injectedClassNames() {
+    return Arrays.asList(
+        "io.opentelemetry.javaagent.instrumentation.kotlinxcoroutines.instrumentationannotations.AnnotationSingletons",
+        "io.opentelemetry.javaagent.instrumentation.kotlinxcoroutines.instrumentationannotations.AnnotationInstrumentationHelper",
+        "io.opentelemetry.javaagent.instrumentation.kotlinxcoroutines.instrumentationannotations.MethodRequest",
+        "io.opentelemetry.javaagent.instrumentation.kotlinxcoroutines.instrumentationannotations.MethodRequestCodeAttributesGetter");
+  }
```

Comment:
instead of injecting all the helper did you consider adding a proxy for `AnnotationInstrumentationHelper` similarly to https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/edbd9aa8ab0ba9588a5fe21eff6ddc821cc0874e/instrumentation/graphql-java/graphql-java-20.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/graphql/v20_0/GraphqlInstrumentationModule.java#L41-L48

Reply:
I think that could work, I will try to see if we can easily do it.

If I understand what you aim to achieve here is only exposing the `AnnotationInstrumentationHelper` class to the instrumented code through an injected proxy, and then leave other classes called by the real `AnnotationInstrumentationHelper` class implementation as implementation details.

One of the downsides of this approach is that we loose the type-safety of the method signatures and need to introduce explicit casts to make it work.

Reply:
> One of the downsides of this approach is that we loose the type-safety of the method signatures and need to introduce explicit casts to make it work.

When method is called from asm we already have lost the type-safety as compiler can't verify that we are passing the correct arguments or whether the method exists at all. Having to use an explicit cast is unfortunate, but since our design is such that the helper types are in a different loader and not visible to instrumented type there isn't an easy way around this. In the future when we stop shading otel api we'll run into the same issue with the context and scope parameters. Perhaps at that point we'll have to rethink this instrumentation a bit, maybe encapsulate request/context/scope into one object.